### PR TITLE
Create an ObservedAddrManager and add an AddressMapper in AutonatService and AutoRelayService

### PR DESCRIPTION
**Pull request by diegomrsantos**
- Creates an `ObservedAddrManager` which keeps track of the most observed addresses reported by remote peers.
- Creates an `AddressMapper` in `AutonatService` which tries to identify the peer public IP/Port when there's a manual port forward configured. 
-  Creates an `AddressMapper` in `AutonRelayService` which adds the relayed address.

### CODE REVIEW

File: libp2p/observedaddrmanager.nim [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    minCount: int
+
+proc `<`(a, b: ObservedAddr): bool = a.count < b.count
+
+proc add(self:ObservedAddrManager, heap: var HeapQueue[ObservedAddr], observedAddr: MultiAddress) =
+  if heap.len >= self.maxSize:
+    discard heap.pop()
+
+  for i in 0 ..< heap.len:
+    if heap[i].ma == observedAddr:
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/observedaddrmanager.nim]: Replaced the heap with a seq.

File: libp2p/peerstore.nim [hunk 2]
```diff
@@ -78,11 +79,13 @@ type
     identify: Identify
     capacity*: int
     toClean*: seq[PeerId]
+    observedAddrManager*: ObservedAddrManager
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/peerstore.nim]: Done.

File: libp2p/observedaddrmanager.nim [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  ## Returns the most observed IP4 address or none if the number of observations are less than minCount.
+  return self.getIP(self.observedIPs, IP4)
+
+proc getMostObservedIP6AndPort*(self: ObservedAddrManager): Opt[MultiAddress] =
+  ## Returns the most observed IP6/Port address or none if the number of observations are less than minCount.
+  return self.getIP(self.observedIPsAndPorts, IP6)
+
+proc getMostObservedIP4AndPort*(self: ObservedAddrManager): Opt[MultiAddress] =
+  ## Returns the most observed IP4/Port address or none if the number of observations are less than minCount.
+  return self.getIP(self.observedIPsAndPorts, IP4)
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/observedaddrmanager.nim]: Did something similar

File: libp2p/protocols/connectivity/autonat/service.nim [hunk 4]
```diff
@@ -153,7 +154,69 @@ proc schedule(service: AutonatService, switch: Switch, interval: Duration) {.asy
...
+  var addrs = newSeq[MultiAddress]()
+  for listenAddr in listenAddrs:
+    try:
+      let maIP = listenAddr[0]
+      if maIP.isErr():
+        continue
+      var isIP4 = true
+      let hostIP =
+        if IP4.match(maIP.get()):
+          getBestRoute(initTAddress("8.8.8.8:0")).source
```
Comment (Menduist) [inline libp2p/protocols/connectivity/autonat/service.nim:199]: So we leave getBestRoute here, instead of in a dedicated service? Seems a bit unrelated to the AutonatService

File: libp2p/protocols/connectivity/autonat/service.nim [hunk 5]
```diff
@@ -153,7 +154,69 @@ proc schedule(service: AutonatService, switch: Switch, interval: Duration) {.asy
...
+          isIP4 = false
+          getBestRoute(initTAddress("2600:::0")).source
+        else:
+          continue
+      if not hostIP.isGlobal():
+        if self.networkReachability == NetworkReachability.Reachable:
+          let newMA = handleManualPortForwarding(peerStore, listenAddr, isIP4)
+          if newMA.isSome():
+            addrs.add(newMA.get())
+          continue
```
Comment (Menduist) [inline libp2p/protocols/connectivity/autonat/service.nim:210]:
```nim
else:
   # replace with hostIp
```

**Review by Menduist — COMMENTED** [review]
We will also have to check what impact this has on the existing addressMapper in waku

File: libp2p/observedaddrmanager.nim [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+type
+  ## Manages observed MultiAddresses by reomte peers. It keeps track of the most observed IP and IP/Port.
+  ObservedAddrManager* = ref object of RootObj
+    observedIPs: seq[MultiAddress]
+    observedIPsAndPorts: seq[MultiAddress]
+    maxSize: int
+    minCount: int
+
+  IPVersion* = enum
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/observedaddrmanager.nim:28]: My reasoning was to not require knowledge of chronos or other unrelated code (like MaPattern) to use this feature.

File: libp2p/observedaddrmanager.nim [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import
+  std/[sequtils, tables],
+  chronos,
+  ./multiaddress
+
+type
+  ## Manages observed MultiAddresses by reomte peers. It keeps track of the most observed IP and IP/Port.
+  ObservedAddrManager* = ref object of RootObj
+    observedIPs: seq[MultiAddress]
+    observedIPsAndPorts: seq[MultiAddress]
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/observedaddrmanager.nim:24]: True, using a seq simplifies many things.

File: libp2p/protocols/connectivity/autonat/service.nim (hunk 4 above)
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/autonat/service.nim:199]: I think you're right, we don't need this anymore in the current version of this code.

File: libp2p/protocols/connectivity/autonat/service.nim (hunk 5 above)
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/autonat/service.nim:210]: Did something different, let me know what you think.

File: tests/testobservedaddrmanager.nim [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+suite "ObservedAddrManager":
+  teardown:
+    checkTrackers()
+
+  asyncTest "Calculate the most oberserved IP correctly":
+
+    let observedAddrManager = ObservedAddrManager.new(minCount = 3)
+
+    # Calculate the most oberserved IP4 correctly
+    let mostObservedIP4AndPort = MultiAddress.init("/ip4/1.2.3.0/tcp/1").get()
```
Comment (Menduist) [inline tests/testobservedaddrmanager.nim:15]: Can you test with invalid multiaddresses, seems like it will crash

File: libp2p/protocols/connectivity/autonat/service.nim [hunk 9]
```diff
@@ -133,6 +133,7 @@ proc askPeer(self: AutonatService, switch: Switch, peerId: PeerId): Future[Netwo
   await self.handleAnswer(ans)
   if not isNil(self.statusAndConfidenceHandler):
     await self.statusAndConfidenceHandler(self.networkReachability, self.confidence)
+  await switch.peerInfo.update()
```
[the remark this answers is not in the document]
Comment (Menduist) [reply libp2p/protocols/connectivity/autonat/service.nim:138]: Up? Can just store the last state and only update if it's !=

File: libp2p/protocols/connectivity/autonat/service.nim [hunk 10]
```diff
@@ -153,7 +156,30 @@ proc schedule(service: AutonatService, switch: Switch, interval: Duration) {.asy
...
+  peerStore: PeerStore,
+  listenAddrs: seq[MultiAddress]): Future[seq[MultiAddress]] {.gcsafe, async.} =
+
+  var addrs = newSeq[MultiAddress]()
+  for listenAddr in listenAddrs:
+    var processedMA = listenAddr
+    try:
+      let hostIP = initTAddress(listenAddr).get()
+      if not hostIP.isGlobal():
+        if self.networkReachability == NetworkReachability.Reachable:
```
Comment (Menduist) [inline libp2p/protocols/connectivity/autonat/service.nim:170]: Maybe this if should be the first thing we do? ie early return if we are not reachable

File: libp2p/protocols/connectivity/autonat/service.nim (hunk 9 above)
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/autonat/service.nim:138]: I think the addresses could still change even if the state hasn't changed.

File: libp2p/protocols/connectivity/autonat/service.nim (hunk 10 above)
Comment (diegomrsantos) [reply libp2p/protocols/connectivity/autonat/service.nim:170]: This is the only thing which is done. I merged both ifs.

File: libp2p/protocols/connectivity/autonat/service.nim (hunk 9 above)
[the remark this answers is not in the document]
Comment (Menduist) [reply libp2p/protocols/connectivity/autonat/service.nim:138]: Fair, though that depends on the observed addresses, not on the autonat service
But hard to model that, I think

File: libp2p/peerstore.nim [hunk 11]
```diff
@@ -220,3 +220,40 @@ proc identify*(
       peerStore.updatePeerInfo(info)
   finally:
     await stream.closeWithEOF()
+
+proc getMostObservedIP*(self: PeerStore, ipVersion: IPVersion): Opt[MultiAddress] =
+  ## Returns the most observed IP address or none if the number of observations are less than minCount.
+  return self.identify.getMostObservedIP(ipVersion)
+
+proc replaceMAIpByMostObserved*(
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/peerstore.nim:228]: Where would you add it? I think it is related to the observed addresses.

[the remark this answers is not in the document]
Comment (Menduist) [reply libp2p/peerstore.nim:228]: ObservedAddressManager?

[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/peerstore.nim]: done
